test/functional/dnsmasq_controller_test.go

```diff
@@ -0,0 +1,306 @@
+		})
+
+		It("stored the input hash in the Status", func() {
+			Eventually(func(g Gomega) {
+				instance := GetDNSMasq(dnsMasqName)
+				g.Expect(instance.Status.Hash).Should(HaveKeyWithValue("input", Not(BeEmpty())))
+			}, timeout, interval).Should(Succeed())
+		})
+
+		It("exposes the service", func() {
+			th.ExpectCondition(
+				dnsMasqName,
+				ConditionGetterFunc(DNSMasqConditionGetter),
+				condition.ExposeServiceReadyCondition,
+				corev1.ConditionTrue,
+			)
+
+			svc := th.GetService(types.NamespacedName{
+				Namespace: namespace,
+				Name:      fmt.Sprintf("dnsmasq-%s-ctlplane", dnsMasqName.Name)})
+			Expect(svc.Labels["service"]).To(Equal("dnsmasq"))
+		})
+
+		It("creates a Deployment for the service", func() {
+
+			th.ExpectConditionWithDetails(
+				dnsMasqName,
+				ConditionGetterFunc(DNSMasqConditionGetter),
+				condition.DeploymentReadyCondition,
+				corev1.ConditionFalse,
+				condition.RequestedReason,
+				condition.DeploymentReadyRunningMessage,
+			)
+
+			deploymentName = types.NamespacedName{
+				Name:      fmt.Sprintf("dnsmasq-%s", dnsMasqName.Name),
+				Namespace: namespace,
+			}
+
+			Eventually(func(g Gomega) {
+				depl := th.GetDeployment(deploymentName)
+
+				g.Expect(int(*depl.Spec.Replicas)).To(Equal(1))
+				g.Expect(depl.Spec.Template.Spec.Volumes).To(HaveLen(3))
+				g.Expect(depl.Spec.Template.Spec.Containers).To(HaveLen(1))
+				g.Expect(depl.Spec.Template.Spec.InitContainers).To(HaveLen(1))
+				g.Expect(depl.Spec.Selector.MatchLabels).To(Equal(map[string]string{"service": "dnsmasq"}))
+
+				container := depl.Spec.Template.Spec.Containers[0]
+				g.Expect(container.VolumeMounts).To(HaveLen(3))
+				g.Expect(container.Image).To(Equal(containerImage))
+
+				g.Expect(container.LivenessProbe.TCPSocket.Port.IntVal).To(Equal(int32(53)))
+				g.Expect(container.ReadinessProbe.TCPSocket.Port.IntVal).To(Equal(int32(53)))
+			}, timeout, interval).Should(Succeed())
+		})
+
+		When("the DNSData CM gets updated", func() {
+			It("the CONFIG_HASH on the deployment changes", func() {
+				cm := th.GetConfigMap(dnsDataCM)
+				configHash := ""
+				deploymentName = types.NamespacedName{
+					Name:      fmt.Sprintf("dnsmasq-%s", dnsMasqName.Name),
+					Namespace: namespace,
+				}
+				Eventually(func(g Gomega) {
+					depl := th.GetDeployment(deploymentName)
+					found := false
+					for _, vol := range depl.Spec.Template.Spec.Volumes {
+						if vol.Name == dnsDataCM.Name {
+							found = true
+						}
+					}
+					g.Expect(found).To(BeTrue())
+					container := depl.Spec.Template.Spec.Containers[0]
+					found = false
+					for _, vol := range container.VolumeMounts {
+						if vol.Name == dnsDataCM.Name {
+							found = true
+						}
+					}
+					g.Expect(found).To(BeTrue())
+					for _, env := range container.Env {
+						if env.Name == "CONFIG_HASH" {
+							configHash = container.Env[0].Value
+							break
+						}
+					}
```

Comment:
Ohh, I think I have a helper for this in https://github.com/openstack-k8s-operators/nova-operator/blob/c170e2201e05ecc8345d6a5f000fc3db0dd7838d/test/functional/base_test.go#L782-L789 . I will move that to lib-common so you can reuse it.

Reply:
Pushed https://github.com/openstack-k8s-operators/lib-common/pull/270 It is OK to adapt to this in a follow up.

Reply:
as it is merged, let me get that in.

Reply:
ack

Reply:
done

Reply:
looks good. thanks.